File: web/src/lib/todos.ts
/**
 * What a todo is, as the two surfaces that draw one both need it.
 *
 * A todo is an artifact of type memory and kind todo, and everything else about
 * it is convention the queue has carried since before there was a view: the
 * status words, the reading order, and the OWNER line at the top of the body.
 * The global page and the room panel have to agree about all three - two ideas
 * of what "active" sorts above, or two ways of finding the owner, is a queue
 * that reads differently depending on where you look at it from.
 */

import type { Artifact } from "@/lib/api";

/**
 * The order statuses are shown in. Active first, then open, then done: a list
 * that buries what is in flight under what is finished answers no question
 * anybody opened it to ask. Everything unrecognised sorts with the open ones -
 * an unknown status is work nobody has said is done.
 */
const RANK: Record<string, number> = { active: 0, todo: 1, done: 2 };

export function todoRank(status: string): number {
  return RANK[status.trim().toLowerCase()] ?? RANK.todo;
}

/** sortTodos puts a queue in reading order, keeping the node's order within a status. */
export function sortTodos(list: Artifact[]): Artifact[] {
  return [...list].sort((a, b) => todoRank(a.status) - todoRank(b.status));
}

/**
 * The kinds of memory the node treats as WORK: the ones with a status, a
 * dependency edge and a note door. It mirrors store.WorkKinds, which is the
 * list the node refuses everything else against.
 *
 * Mirrored rather than asked for, on the same terms as TODO_KINDS below: the
 * set is closed and changes about once a year, and the alternative is a fetch
 * before the page can decide whether to draw a box. What the mirror must never
 * do is decide anything the node then contradicts, which is why it is used only
 * to choose whether to OFFER the note box - a row this list gets wrong shows a
 * box whose write the node refuses in words, rather than a refusal invented
 * over here.
 */
const WORK_KINDS = ["todo", "feature", "handoff", "merge", "work"];

/**
 * isQueueItem says a row is one of those: a memory of a work kind. It is what
 * the notes section is drawn for, because a report or a proposal has no note
 * door and offering one on it would be a control the node answers with a 404.
 */
export function isQueueItem(artifact: Artifact): boolean {
  return artifact.type === "memory" && WORK_KINDS.includes((artifact.kind ?? "").trim());
}

/** isTodoDone says a piece of work is finished, by the same ranking the list sorts by. */
export function isTodoDone(todo: Artifact): boolean {
  return todoRank(todo.status) === RANK.done;
}

/**
 * Whether the panel draws the finished ones, remembered across reloads.
 *
 * #general holds 26 todos and 16 of them are done, in a panel beside a
 * conversation with about fifteen visible rows - so the finished work pushes the
 * live work off the bottom, and a panel that exists to answer "what is this room
 * doing" mostly answers "what has this room finished".
 *
 * DEFAULT HIDDEN. The other choice keeps today's behaviour and fixes nothing for
 * anybody who never finds the checkbox, which is most people; a default nobody
 * discovers is a feature nobody has. What keeps hiding honest rather than lossy
 * is that the number hidden is on screen the whole time it is hiding anything,
 * beside the box that did it - a panel showing four rows and no hint that
 * sixteen are behind it is how somebody concludes a todo does not exist.
 *
 * ONE setting for every room rather than one key per room. This is a habit about
 * reading a panel, not a fact about a room: somebody who wants finished work out
 * of the way in #general wants it out of the way in #build too, and a per-room
 * key means ticking the same box once per room and finding the next one buried
 * again. It also keeps the panel honest across a room switch - the component
 * stays mounted when the room changes, so a per-room value would have to be
 * re-read on that change or the new room would be drawn under the old room's
 * setting.
 */
const HIDE_DONE_KEY = "flowy.todos.hideDone";

export function hideDonePreference(): boolean {
  try {
    // Anything that is not an explicit "false" is hidden, so a browser with
    // nothing stored yet gets the default.
    return localStorage.getItem(HIDE_DONE_KEY) !== "false";
  } catch {
    return true;
  }
}

export function setHideDonePreference(hide: boolean) {
  try {
    localStorage.setItem(HIDE_DONE_KEY, hide ? "true" : "false");
  } catch {
    // Storage switched off. The setting still holds for the length of the page,
    // the same way the token does.
  }
}

/** countTodos is the header line: how many are in flight, waiting, and finished. */
export function countTodos(list: Artifact[]): { active: number; open: number; done: number } {
  const counts = { active: 0, open: 0, done: 0 };
  for (const t of list) {
    if (todoRank(t.status) === RANK.active) counts.active += 1;
    else if (todoRank(t.status) === RANK.done) counts.done += 1;
    else counts.open += 1;
  }
  return counts;
}

/**
 * The owner is the first line of the body, not the artifact's owner_user.
 *
 * owner_user is whoever wrote the row, which for most of this queue is the one
 * operator principal that filed all of it - it says "operator" and answers
 * nothing. The body carries `OWNER: <name>` as its first line, which is the
 * claim somebody actually made about who is doing the work, and a literal `?`
 * there means nobody has taken it.
 */
export function todoOwner(body: string): string {
  const line = body.split("\n").find((l) => l.startsWith("OWNER:"));
  const name = line?.slice("OWNER:".length).trim();
  if (!name || NOBODY.has(name.toLowerCase())) return "";
  return name;
}

/**
 * The words that mean nobody is carrying this. They all collapse to the empty
 * owner, so the panel says ONE word for one state.
 *
 * Raised as a todo through the panel itself: 'todo list has "unowned" and
 * "unassigned" - looks identical. triage and fix'. Both were there because the
 * render falls back to "unowned" while several bodies had been written with
 * "OWNER: unassigned" - two words for one state, which reads as two states and
 * makes a reader look for a distinction that does not exist.
 *
 * Normalising here rather than rewriting those bodies is the durable half: the
 * next person to write "OWNER: none" or "OWNER: TBD" gets the same single word,
 * without anybody having to know the convention.
 */
const NOBODY = new Set(["?", "-", "none", "nobody", "tbd", "unassigned", "unowned", "n/a"]);

/**
 * Who is carrying a todo: the `assignee` field if the item has one, and the
 * body's OWNER line if it does not.
 *
 * The order is the compatibility, and it is the same order the node and the
 * terminal client read these in. The whole queue predates the field, and those
 * items still read the way they always did. But a field that is THERE wins even
 * when it is empty - somebody who unassigned a todo through this panel said so
 * out loud, and falling through to the OWNER line still sitting in the body
 * would quietly put the old name back on the next render.
 *
 * So the presence of the key is the question, not its truthiness: `""` is a
 * value here and `undefined` is a silence, which is why this cannot go through
 * fieldOf below.
 */
export function todoAssignee(artifact: Artifact): string {
  const fields = artifact.fields;
  if (fields && typeof fields === "object") {
    const named = (fields as Record<string, unknown>).assignee;
    if (typeof named === "string") return named.trim();
  }
  return todoOwner(artifact.body ?? "");
}

/**
 * WHO RAISED IT: who the work came from, beside who is carrying it.
 *
 * These are two facts and the row draws both, because a row that draws one of
 * them is ambiguous in exactly the way this field exists to fix. Four agents
 * share this board and one of them files rows out of conversations the operator
 * had - `owner_user` is then the agent, which is true and is not where the work
 * came from, and nothing on the row said so.
 *
 * It is NOT `owner_user` and it is not a fallback for it. owner_user is the seat
 * whose token wrote the row - the signing author, and the answer to a different
 * question - so a todo with no raiser is drawn with none rather than with the
 * author's id standing in for one. Most of this queue was written before the
 * field and says nothing here, which is the truth about it.
 *
 * Read off `fields`, like the room and the message beside it: the node also
 * derives it onto the row at read time, and one place to dig is what keeps the
 * two clients agreeing about what an absent one means.
 */
export function todoRaiser(artifact: Artifact): string {
  return fieldOf(artifact, "raiser").trim();
}

/**
 * THE KIND OF WORK, which the node calls `category`.
 *
 * Two different things are labelled on a todo and they are not variants of each
 * other. TAGS are free labels: any number of them, any word, nobody's schema,
 * and they are searched with the title and the body. THIS is one value out of a
 * closed set the node REFUSES anything outside of - which is the only reason it
 * can be counted or routed on. A filter over tags answers "whatever people
 * typed"; a filter over this answers "the bugs".
 *
 * It is called "Kind" here and `category` on the wire, and the difference is not
 * an accident. A todo already IS kind=todo one level up, so the node cannot use
 * that word twice without the two meanings being told apart by context - which
 * is exactly the confusion this console should not pass on to the person reading
 * it. They see one word, in the place a person expects it; the wire keeps a word
 * that can only mean one thing.
 *
 * Empty is a todo nobody has classified. That is legal, it is most of this
 * queue, and it is drawn as such rather than being guessed at from the title.
 */
export const TODO_KINDS = ["bug", "feature", "chore", "question"] as const;

export type TodoKind = (typeof TODO_KINDS)[number];

/** What the node calls it, off fields. Unknown words read as unclassified: the
 * node refuses them on the way in, so one here came from somewhere this console
 * cannot fix and is not a word to paint as though it were in the set. */
export function todoKind(artifact: Artifact): string {
  const named = fieldOf(artifact, "category").trim().toLowerCase();
  return (TODO_KINDS as readonly string[]).includes(named) ? named : "";
}

/**
 * The colour a kind is drawn in, on the same terms as statusStyle: the word
 * stays inside the badge, so colour is the second signal and never the only
 * one, and the colours are picked for what each one means.
 *
 * RED for a bug, because it is the one an operator scans for. BLUE for a
 * feature, GREY-GREEN for a chore - work that has to happen and is nobody's
 * news - and VIOLET for a question, which is the one that is not yet a piece of
 * work at all.
 */
export function kindStyle(kind: string): { color: string; backgroundColor: string } {
  const colour = KIND_COLOUR[kind] ?? "#8b93a7";
  return { color: colour, backgroundColor: `color-mix(in srgb, ${colour} 18%, transparent)` };
}

const KIND_COLOUR: Record<string, string> = {
  bug: "#d1626b", // red - something is broken
  feature: "#5b8dd6", // blue - something new
  chore: "#7d9c8a", // quiet green - has to happen, is not news
  question: "#a481d4", // violet - not yet a piece of work
};

/**
 * The free labels on an item: the author's tags and the node's own, as one
 * list, deduplicated and in a stable order.
 *
 * Both columns are shown because both are labels a person put there - user_tags
 * is what somebody typed on their own item and tags is what the surfaces that
 * write items fill in - and a console that drew one of them would be hiding half
 * the labels behind a distinction nobody reading the queue is making.
 */
export function todoTags(artifact: Artifact): string[] {
  const all = [...(artifact.tags ?? []), ...(artifact.user_tags ?? [])];
  const seen = new Set<string>();
  const out: string[] = [];
  for (const raw of all) {
    const tag = raw.trim();
    if (!tag || seen.has(tag)) continue;
    seen.add(tag);
    out.push(tag);
  }
  return out.sort((a, b) => a.localeCompare(b));
}

/** Every tag in a queue, for the filter to offer. It is built from the rows on
 * the page rather than from a list somewhere: tags have no schema, so what
 * exists is whatever has been written. */
export function tagsIn(list: Artifact[]): string[] {
  const seen = new Set<string>();
  for (const todo of list) {
    for (const tag of todoTags(todo)) seen.add(tag);
  }
  return [...seen].sort((a, b) => a.localeCompare(b));
}

/** The body without the OWNER line, which is rendered as its own column. */
export function todoDetail(body: string): string {
  return body
    .split("\n")
    .filter((l) => !l.startsWith("OWNER:"))
    .join("\n")
    .trim();
}

/**
 * Where a todo was raised, off fields. The node keeps `room` and `message`
 * there - fields is jsonb of whatever the type puts in it, so it is narrowed
 * here at the use site rather than typed on Artifact for every type at once.
 */
function fieldOf(artifact: Artifact, key: string): string {
  const fields = artifact.fields;
  if (!fields || typeof fields !== "object") return "";
  const value = (fields as Record<string, unknown>)[key];
  return typeof value === "string" ? value : "";
}

export function todoRoom(artifact: Artifact): string {
  return fieldOf(artifact, "room");
}

/** The message the todo was raised out of, when it was raised in a room. */
export function todoMessage(artifact: Artifact): string {
  return fieldOf(artifact, "message");
}

/**
 * The colour a status is drawn in. Asked for directly: "I wanted colors for
 * Active Done and Todo".
 *
 * Three states, three jobs, and the colours are picked for what each one means
 * rather than for variety. AMBER for active, because in-flight work is the
 * thing you want your eye to land on when you open the panel. GREEN for done,
 * which is the one convention here strong enough to be worth obeying. GREY for
 * todo, deliberately quiet: a queue is mostly waiting, and if waiting shouts
 * then nothing does.
 *
 * The status word stays inside the badge. Colour on its own would leave anybody
 * who cannot separate amber from green reading a queue with no states in it,
 * and this is a panel about what is happening rather than a decoration.
 *
 * Explicit colours rather than theme tokens, for the same reason speaker
 * colours are: a status is a fact about the work, not a role in the interface,
 * and "primary" and "destructive" already mean other things here.
 */
export function statusStyle(status: string): { color: string; backgroundColor: string } {
  const colour = STATUS_COLOUR[todoRank(status)] ?? STATUS_COLOUR[RANK.todo];
  return { color: colour, backgroundColor: `color-mix(in srgb, ${colour} 18%, transparent)` };
}

const STATUS_COLOUR: Record<number, string> = {
  [RANK.active]: "#e0a03f", // amber - in flight, and the first thing to see
  [RANK.todo]: "#8b93a7", // grey - waiting, and quiet on purpose
  [RANK.done]: "#4fae7a", // green - finished
};

File: internal/tui/todos.go
package tui

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// The todos view: the shared queue, who owns each item, and what it is waiting
// on.
//
// It exists because the queue was real and invisible. Four agents and a person
// working in one room had the state of the work in chat messages and summaries,
// which is to say nowhere: what is in flight, who has it, and what is blocked
// behind what were things you found out by asking. The items themselves have
// been artifacts for a while - type memory, kind todo, project scope - and
// nothing rendered them, so the fabric knew the answer and no surface said it.
//
// The ordering is the feature and not a preference. Active first, then the
// open ones, then done: a list that buries what is in flight under what is
// finished answers none of the three questions somebody opens this to ask. Done
// is kept rather than filtered because "is that finished?" is one of them.
//
// It reads and does not write. Closing a todo is a lifecycle question - what
// done means here, who may say it, and what the trail records - and inventing
// an answer to that on a key in a terminal is how a queue ends up with two
// disagreeing ideas of its own state. mem_write already has one.

// todoStatuses is what an item's status may say, in the order the list puts
// them in. Anything else the queue grows later sorts with the open ones rather
// than with the finished ones: an unrecognised status is work nobody has said
// is done.
const (
	todoActive = "active"
	todoDone   = "done"
	todoTodo   = "todo"
)

// todoRank is where a status sorts. The gap at 2 is deliberate - it is where
// every status this client has never heard of lands, above done and below
// nothing.
func todoRank(status string) int {
	switch strings.TrimSpace(strings.ToLower(status)) {
	case todoActive:
		return 0
	case todoTodo, "":
		return 1
	case todoDone:
		return 3
	default:
		return 2
	}
}

// todoStyle is how a status is drawn, which is the same three meanings the
// console draws them with - see statusStyle in web/src/lib/todos.ts.
//
// The colour is the second half of the row and never the whole of it. A
// terminal may be monochrome, piped into less, or read by somebody who cannot
// separate amber from green, and lipgloss quietly renders all three of these as
// plain text on a TERM=dumb - so the status WORD goes on the row as well, every
// time. A queue whose states are only a colour has no states at all for the
// reader who cannot see them.
func (t Theme) todoStyle(status string) lipgloss.Style {
	switch todoRank(status) {
	case 0:
		return t.TodoActive
	case 3:
		return t.TodoDone
	default:
		// Including the statuses this client has never heard of, which sort with
		// the open ones and are drawn as them: unrecognised is not finished.
		return t.TodoOpen
	}
}

// sortTodos puts the queue in reading order: active, open, done.
//
// It is a stable sort, so within a status the node's own order survives - which
// is what it came back in, and the only ordering the node has an opinion about.
func sortTodos(list []*Artifact) []*Artifact {
	sort.SliceStable(list, func(i, j int) bool {
		return todoRank(list[i].Status) < todoRank(list[j].Status)
	})
	return list
}

// todoOwner is who is doing it: the assignee field if the item carries one, and
// the OWNER line the queue's older items start with if it does not.
//
// Neither is owner_user, which is the ulid of whoever wrote the row - usually
// the one agent that filed the whole queue, and never the answer to "who is
// doing this". Items that say neither render with a dash: an unowned todo is a
// fact about the queue worth seeing, not a row to hide.
//
// The field wins even when it is empty, which is the same order the console and
// the node read these in. Somebody who unassigned a todo through a panel said
// so; falling back to the OWNER line still in its body would undo them here and
// leave the two clients disagreeing about who has the work.
func todoOwner(a *Artifact) string {
	if a == nil {
		return ""
	}
	if named := todoProvenance(a).Assignee; named != nil {
		return somebody(*named)
	}
	for _, line := range splitLines(a.Body) {
		line = strings.TrimSpace(line)
		if rest, ok := strings.CutPrefix(line, "OWNER:"); ok {
			return somebody(rest)
		}
		if line != "" {
			// The convention is the first line of the body. Scanning the whole
			// body would find the word in a sentence about somebody else's item.
			return ""
		}
	}
	return ""
}

// nobodyWords are the ways this queue has said that nobody is carrying an item.
// They all collapse to the empty owner, so every surface says ONE word for one
// state.
//
// It was raised as a todo through the panel itself: 'todo list has "unowned"
// and "unassigned" - looks identical. triage and fix'. Two words for one state
// read as two states and send a reader looking for a distinction that is not
// there. The node normalises what it is handed (nobodyWords in todos.go) and
// the console collapses the same set on the way out (web/src/lib/todos.ts);
// this is the third copy, for the bodies written before either existed.
var nobodyWords = map[string]bool{
	"?": true, "-": true, "none": true, "nobody": true,
	"tbd": true, "unassigned": true, "unowned": true, "n/a": true,
}

// somebody is a name that means somebody, or "" for one of the words above.
func somebody(name string) string {
	name = strings.TrimSpace(name)
	if nobodyWords[strings.ToLower(name)] {
		return ""
	}
	return name
}

// todoFields is what a todo carries in fields: where it was raised, the message
// it came out of, and who is carrying it. It rides fields rather than columns
// for the reason a report's as_of does - see reportProvenance, which is the same
// read of the same column - and anything that does not parse leaves them empty,
// which is a todo with no room on screen and is the truth.
//
// Assignee is a pointer because an empty one is a value: "nobody is carrying
// this" is a thing somebody said, and it has to be told apart from an item that
// predates the field and whose owner is still a line in its body.
type todoFields struct {
	Room     string  `json:"room"`
	Message  string  `json:"message"`
	Assignee *string `json:"assignee"`
	// Raiser is who the work came from, and it is a plain string where the
	// assignee is a pointer, because the two absences are not the same absence.
	// An empty assignee is somebody saying nobody is carrying this; a raiser is
	// settled when the row is raised and never restated, so an absent one is a
	// row that does not say where its work came from - which is every queue item
	// written before the field, and is not a claim anybody made.
	Raiser string `json:"raiser"`
}

func todoProvenance(a *Artifact) todoFields {
	var fields todoFields
	if a == nil || len(a.Fields) == 0 {
		return fields
	}
	_ = json.Unmarshal(a.Fields, &fields)
	return fields
}

// todoRoomOf is the room a todo was raised in, or "" for one raised in none.
func todoRoomOf(a *Artifact) string { return todoProvenance(a).Room }

// todoRaiser is who the work came from: the party that asked for it, which is
// not the party carrying it and is not owner_user either.
//
// owner_user is the seat whose token wrote the row - one agent for a whole
// board it filed on somebody else's behalf - so this client has never drawn it
// and does not start here. A row that says nothing gets nothing: there is no
// older convention to fall back on the way todoOwner falls back on an OWNER
// line, and guessing the author would put a name on the screen that nobody
// claimed.
func todoRaiser(a *Artifact) string { return somebody(todoProvenance(a).Raiser) }

// todoCounts is the header: how many are in flight, how many are waiting, and
// how many are finished.
func todoCounts(list []*Artifact) (active, open, done int) {
	for _, a := range list {
		switch todoRank(a.Status) {
		case 0:
			active++
		case 3:
			done++
		default:
			open++
		}
	}
	return active, open, done
}

func (m *Model) todosKey(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "/":
		return m, m.openInput(inputTodoQuery, "search> ", m.todoQuery)
	case "enter":
		if selected := m.selectedTodo(); selected != nil {
			m.backView = viewTodos
			m.view = viewArtifact
			return m, m.artifactCmd(selected.ID)
		}
		return m, nil
	case "c":
		m.todoQuery = ""
		return m, m.todosCmd("")
	}
	if sel, moved := moveSel(msg.String(), m.todoSel, len(m.todos), m.height/2); moved {
		m.todoSel = sel
	}
	return m, nil
}

func (m *Model) selectedTodo() *Artifact {
	if m.todoSel < 0 || m.todoSel >= len(m.todos) {
		return nil
	}
	return m.todos[m.todoSel]
}

// todoOwnerWidth is what each of the two name columns gets. A handle has no
// length limit and a terminal does: at 80 columns the status and the two names
// take 37 of them and the title keeps the rest, which is the point of fixing
// them rather than letting one long name push what the item is off the right of
// the row.
const todoOwnerWidth = 14

// todoStatusColumn is what the status word gets, and it is the width the header
// below lines its own labels up on.
const todoStatusColumn = 8

// todoHeader names the columns, and it is here because the row grew a SECOND
// name.
//
// One name needed no label - a queue row was a state and a person, and which
// person was obvious. Two are not: "raised by" and "carried by" are different
// claims and are frequently different parties, and a reader who has to work out
// which column is which from the values in it will get it wrong on the rows
// where it matters, which are the rows where they differ. The console says the
// words on every row because it has the width for it; a terminal says them once
// at the top.
func todoHeader() string {
	return fmt.Sprintf("%-*s %-*s %-*s %s",
		todoStatusColumn, "status", todoOwnerWidth, "raised by",
		todoOwnerWidth, "carried by", "what")
}

func (m *Model) todosView(height int) []string {
	active, open, done := todoCounts(m.todos)
	head := fmt.Sprintf("todos (%d active, %d open, %d done)", active, open, done)
	if m.todoQuery != "" {
		head = fmt.Sprintf("todos matching %q (%d active, %d open, %d done)",
			m.todoQuery, active, open, done)
	}
	lines := []string{m.theme.Title.Render(m.theme.clip(head, m.width))}
	if len(m.todos) == 0 {
		// Where they come from, because this pane cannot make one and an empty
		// list otherwise reads as a broken view rather than an empty queue.
		what := "no todos yet - written with mem_write, kind todo"
		if m.todoQuery != "" {
			what = "nothing matched - c clears the search"
		}
		lines = append(lines, m.theme.Dim.Render(m.theme.clip(what, m.width)))
		return lines
	}

	// The same split the memory and reports panes use: the list on top, the
	// selected item under the rule. What is under the rule here is the body,
	// which is where the DEPENDS ON line lives - the half of the question this
	// view was asked for that a row has no room for.
	//
	// Two lines are spent before the list now rather than one - the count and
	// the column labels - so the list gets one fewer.
	lines = append(lines, m.theme.Dim.Render(m.theme.clip(todoHeader(), m.width)))
	listHeight := max(1, (height-2)/2)
	start, end := window(len(m.todos), m.todoSel, listHeight)
	for i := start; i < end; i++ {
		a := m.todos[i]
		owner := todoOwner(a)
		if owner == "" {
			owner = "-"
		}
		// One dash for a row that does not say where the work came from, which
		// is the same dash an unowned row gets and says the same thing: nobody
		// stated this. It is not owner_user standing in - see todoRaiser.
		raiser := todoRaiser(a)
		if raiser == "" {
			raiser = "-"
		}
		status := a.Status
		if status == "" {
			status = todoTodo
		}
		text := fmt.Sprintf("%-*s %-*s %-*s %s",
			todoStatusColumn, m.theme.clip(status, todoStatusColumn),
			todoOwnerWidth, m.theme.clip(raiser, todoOwnerWidth),
			todoOwnerWidth, m.theme.clip(owner, todoOwnerWidth), a.Title)
		text = m.theme.clip(text, m.width)
		if i == m.todoSel {
			lines = append(lines, m.theme.Selected.Render(pad(text, m.width)))
			continue
		}
		// The same three states the room's panel draws, because it is the same
		// queue: a reader who learned the colours in one of the two views has
		// learned them in the other.
		lines = append(lines, m.theme.todoStyle(status).Render(text))
	}

	if selected := m.selectedTodo(); selected != nil && len(lines) < height {
		lines = append(lines, m.theme.Dim.Render(m.theme.clip(m.rule(), m.width)))
		lines = append(lines, m.theme.Badge.Render(m.theme.clip(m.badges(selected), m.width)))
		for _, line := range splitLines(selected.Body) {
			if len(lines) >= height {
				break
			}
			lines = append(lines, m.theme.clip(line, m.width))
		}
	}
	return lines
}
